File: backend/modules/llm/domain/entity/runtime.go
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0

package entity

import (
	"fmt"
	"strconv"
	"time"
)

type Message struct {
	Role             Role   `json:"role"`
	Content          string `json:"content"`
	ReasoningContent string `json:"reasoning_content"`

	// if MultiModalContent is not empty, use this instead of Content
	// if MultiModalContent is empty, use Content
	MultiModalContent []*ChatMessagePart `json:"multi_content,omitempty"`

	Name string `json:"name,omitempty"`

	// only for AssistantMessage
	ToolCalls []*ToolCall `json:"tool_calls,omitempty"`

	// only for ToolMessage
	ToolCallID string `json:"tool_call_id,omitempty"`

	ResponseMeta *ResponseMeta `json:"response_meta,omitempty"`
}

func (m *Message) GetInputToken() int {
	if m == nil || m.ResponseMeta == nil || m.ResponseMeta.Usage == nil {
		return 0
	}
	return m.ResponseMeta.Usage.PromptTokens
}

func (m *Message) GetOutputToken() int {
	if m == nil || m.ResponseMeta == nil || m.ResponseMeta.Usage == nil {
		return 0
	}
	return m.ResponseMeta.Usage.CompletionTokens
}

func (m *Message) HasMultiModalContent() bool {
	if m == nil || len(m.MultiModalContent) == 0 {
		return false
	}
	for _, p := range m.MultiModalContent {
		if p.IsMultiModal() {
			return true
		}
	}
	return false
}

func (m *Message) GetImageCountAndMaxSize() (hasUrl, hasBinary bool, cnt int64, maxSizeInByte int64) {
	if !m.HasMultiModalContent() {
		return hasUrl, hasBinary, cnt, maxSizeInByte
	}
	for _, p := range m.MultiModalContent {
		if p.IsURL() {
			hasUrl = true
			cnt++
			continue
		}
		if p.IsBinary() {
			hasBinary = true
			cnt++
			if maxSizeInByte < int64(len(p.ImageURL.URL)*3/4) {
				maxSizeInByte = int64(len(p.ImageURL.URL) * 3 / 4)
			}
		}
	}
	return hasUrl, hasBinary, cnt, maxSizeInByte
}

type Role string

const (
	// Assistant is the role of an assistant, means the message is returned by ChatModel.
	RoleAssistant Role = "assistant"
	// User is the role of a user, means the message is a user message.
	RoleUser Role = "user"
	// System is the role of a system, means the message is a system message.
	RoleSystem Role = "system"
	// Tool is the role of a tool, means the message is a tool call output.
	RoleTool Role = "tool"
)

type ChatMessagePart struct {
	Type     ChatMessagePartType  `json:"type"`
	Text     string               `json:"text"`
	ImageURL *ChatMessageImageURL `json:"image_url"`
}

func (p *ChatMessagePart) IsMultiModal() bool {
	if p == nil {
		return false
	}
	return p.Type != ChatMessagePartTypeText
}

type ChatMessagePartType string

const (
	// ChatMessagePartTypeText means the part is a text.
	ChatMessagePartTypeText ChatMessagePartType = "text"
	// ChatMessagePartTypeImageURL means the part is an image url.
	ChatMessagePartTypeImageURL ChatMessagePartType = "image_url"
	// ChatMessagePartTypeAudioURL means the part is an audio url.
	ChatMessagePartTypeAudioURL ChatMessagePartType = "audio_url"
	// ChatMessagePartTypeVideoURL means the part is a video url.
	ChatMessagePartTypeVideoURL ChatMessagePartType = "video_url"
	// ChatMessagePartTypeFileURL means the part is a file url.
	ChatMessagePartTypeFileURL ChatMessagePartType = "file_url"
)

type ChatMessageImageURL struct {
	// URL can either be a traditional URL or a special URL conforming to RFC-2397 (https://www.rfc-editor.org/rfc/rfc2397).
	// double check with model implementations for detailed instructions on how to use this.
	URL string `json:"url,omitempty"`
	URI string `json:"uri,omitempty"`
	// Detail is the quality of the image url.
	Detail ImageURLDetail `json:"detail,omitempty"`

	// MIMEType is the mime type of the image, eg. "image/png".
	MIMEType string `json:"mime_type,omitempty"`
}

// ImageURLDetail is the detail of the image url.
type ImageURLDetail string

const (
	// ImageURLDetailHigh means the high quality image url.
	ImageURLDetailHigh ImageURLDetail = "high"
	// ImageURLDetailLow means the low quality image url.
	ImageURLDetailLow ImageURLDetail = "low"
	// ImageURLDetailAuto means the auto quality image url.
	ImageURLDetailAuto ImageURLDetail = "auto"
)

type ToolCall struct {
	// Index is used when there are multiple tool calls in a message.
	// In stream mode, it's used to identify the chunk of the tool call for merging.
	Index *int64 `json:"index,omitempty"`
	// ID is the id of the tool call, it can be used to identify the specific tool call.
	ID string `json:"id"`
	// Type is the type of the tool call, default is "function".
	Type string `json:"type"`
	// Function is the function call to be made.
	Function *FunctionCall `json:"function"`

	// Extra is used to store extra information for the tool call.
	Extra map[string]any `json:"extra,omitempty"`
}

// FunctionCall is the function call in a message.
// It's used in Assistant Message.
type FunctionCall struct {
	// Name is the name of the function to call, it can be used to identify the specific function.
	Name string `json:"name,omitempty"`
	// Arguments is the arguments to call the function with, in JSON format.
	Arguments string `json:"arguments,omitempty"`
}

// ResponseMeta collects meta information about a chat response.
type ResponseMeta struct {
	// FinishReason is the reason why the chat response is finished.
	// It's usually "stop", "length", "tool_calls", "content_filter", "null". This is defined by chat model implementation.
	FinishReason string `json:"finish_reason,omitempty"`
	// Usage is the token usage of the chat response, whether usage exists depends on whether the chat model implementation returns.
	Usage *TokenUsage `json:"usage,omitempty"`
}

// TokenUsage Represents the token usage of chat model request.
type TokenUsage struct {
	// PromptTokens is the number of tokens in the prompt.
	PromptTokens int `json:"prompt_tokens"`
	// CompletionTokens is the number of tokens in the completion.
	CompletionTokens int `json:"completion_tokens"`
	// TotalTokens is the total number of tokens in the request.
	TotalTokens int `json:"total_tokens"`
}

type ToolChoice string

const (
	ToolChoiceAuto     ToolChoice = "auto"
	ToolChoiceRequired ToolChoice = "required"
	ToolChoiceNone     ToolChoice = "none"
)

type ToolInfo struct {
	// The unique name of the tool that clearly communicates its purpose.
	Name string
	// Used to tell the model how/when/why to use the tool.
	// You can provide few-shot examples as a part of the description.
	Desc string

	ToolDefType ToolDefType

	Def string
}

type ToolDefType string

const (
	ToolDefTypeOpenAPIV3 ToolDefType = "open_api_v3"
)

type StreamRespParseResult struct {
	RespMsgs          []*Message
	LastRespMsg       *Message
	ReasoningDuration time.Duration
	FirstTokenLatency time.Duration
}

func (p *ChatMessagePart) IsURL() bool {
	if p == nil || p.Type != ChatMessagePartTypeImageURL || p.ImageURL == nil ||
		p.ImageURL.MIMEType != "" {
		return false
	}
	return true
}

func (p *ChatMessagePart) IsBinary() bool {
	if p == nil || p.Type != ChatMessagePartTypeImageURL || p.ImageURL == nil ||
		p.ImageURL.MIMEType == "" {
		return false
	}
	return true
}

type ResponseFormatType string

const (
	ResponseFormatTypeText ResponseFormatType = "text"
	ResponseFormatTypeJSON ResponseFormatType = "json_object"
)

type ResponseFormat struct {
	Type ResponseFormatType `json:"type,omitempty"`
}

type ParamValue struct {
	Name      string    `json:"name"`
	ParamType ParamType `json:"param_type"`
	Value     string    `json:"value"`
	JsonPath  string    `json:"json_path"`
}

func (p *ParamValue) GetValue() (any, error) {
	switch p.ParamType {
	case ParamTypeBoolean:
		return strconv.ParseBool(p.Value)
	case ParamTypeFloat:
		return strconv.ParseFloat(p.Value, 64)
	case ParamTypeInt:
		return strconv.ParseInt(p.Value, 10, 64)
	case ParamTypeString:
		return p.Value, nil
	default:
		return nil, fmt.Errorf("unsupported param type: %s", p.ParamType)
	}
}

File: frontend/packages/loop-base/api-schema/src/api/idl/llm/domain/runtime.ts
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0
import * as manage from './manage';
export { manage };
import * as common from './common';
export { common };
export interface ModelConfig {
  /** 模型id */
  model_id: string,
  temperature?: number,
  max_tokens?: string,
  top_p?: number,
  stop?: string[],
  tool_choice?: ToolChoice,
  /** support json */
  response_format?: ResponseFormat,
  top_k?: number,
  presence_penalty?: number,
  frequency_penalty?: number,
  /** 与ParamSchema对应 */
  param_config_values?: ParamConfigValue[],
}
export interface ParamConfigValue {
  /** 传给下游模型的key，与ParamSchema.name对齐 */
  name?: string,
  /** 展示名称 */
  label?: string,
  /** 传给下游模型的value，与ParamSchema.options对齐 */
  value?: manage.ParamOption,
}
export interface Message {
  role: Role,
  content?: string,
  multimodal_contents?: ChatMessagePart[],
  /** only for AssistantMessage */
  tool_calls?: ToolCall[],
  /** only for ToolMessage */
  tool_call_id?: string,
  /** collects meta information about a chat response */
  response_meta?: ResponseMeta,
  /**
   * only for AssistantMessage, And when reasoning_content is not empty, content must be empty
   * 8: optional map<string,string> extra
  */
  reasoning_content?: string,
}
export interface ChatMessagePart {
  type?: ChatMessagePartType,
  text?: string,
  image_url?: ChatMessageImageURL,
  /**
   * 4: optional ChatMessageAudioURL audio_url 占位,暂不支持
   * 6: optional ChatMessageFileURL file_url 占位,暂不支持
  */
  video_url?: ChatMessageVideoURL,
}
export interface ChatMessageVideoURL {
  url?: string,
  detail?: VideoURLDetail,
  mime_type?: string,
}
export interface VideoURLDetail {
  fps?: number
}
export interface ChatMessageImageURL {
  url?: string,
  detail?: ImageURLDetail,
  mime_type?: string,
}
export interface ToolCall {
  index?: string,
  id?: string,
  type?: ToolType,
  function_call?: FunctionCall,
}
export interface FunctionCall {
  name?: string,
  arguments?: string,
}
export interface ResponseMeta {
  finish_reason?: string,
  /** 3: optional LogProbs log_probs */
  usage?: TokenUsage,
}
export interface TokenUsage {
  prompt_tokens?: string,
  completion_tokens?: string,
  total_tokens?: string,
}
export interface Tool {
  name?: string,
  desc?: string,
  def_type?: ToolDefType,
  /** 必须使用openapi3.Schema序列化后的json */
  def?: string,
}
export interface BizParam {
  workspace_id?: string,
  user_id?: string,
  /** 使用场景 */
  scenario?: common.Scenario,
  /** 场景实体id(非必填) */
  scenario_entity_id?: string,
  /** 场景实体version(非必填) */
  scenario_entity_version?: string,
  /** 场景实体key(非必填), prompt场景需要传prompt key */
  scenario_entity_key?: string,
}
export interface ResponseFormat {
  type?: ResponseFormatType
}
export type ResponseFormatType = string;
export const response_format_json_object = "json_object";
export const response_format_text = "text";
export enum ToolChoice {
  tool_choice_auto = "auto",
  tool_choice_required = "required",
  tool_choice_none = "none",
}
export enum ToolDefType {
  tool_def_type_open_api_v3 = "open_api_v3",
}
export enum Role {
  role_system = "system",
  role_assistant = "assistant",
  role_user = "user",
  role_tool = "tool",
}
export enum ToolType {
  tool_type_function = "function",
}
export enum ChatMessagePartType {
  chat_message_part_type_text = "text",
  chat_message_part_type_image_url = "image_url",
  /** const ChatMessagePartType chat_message_part_type_audio_url = "audio_url" */
  chat_message_part_type_video_url = "video_url",
}
/** const ChatMessagePartType chat_message_part_type_file_url = "file_url" */
export enum ImageURLDetail {
  image_url_detail_auto = "auto",
  image_url_detail_low = "low",
  image_url_detail_high = "high",
}
export enum MimeTypePrefix {
  mime_prefix_image = "image/",
  mime_prefix_video = "video/",
}